### PR TITLE
Consistency operation failed status management in LRO 

### LINKED ISSUE #11473: [Design] [FormRecognizer] Handle operation failed status consistently across Recognize and Train LROs
## Summary

When a training operation fails, no exception is thrown. The developer needs to manually check their model status and errors to assert it has succeeded before actually making use of the model.

```cs
var op = await trainingClient.StartTrainingAsync(...); // POST request succeeds
await op.WaitForCompletionAsync(); // GET request succeeds and returns a failed operation

// op.Value.Status must be checked.
```

Getting no exceptions could be misinterpreted as a sign that the training operation has succeeded.

However, throwing an exception in case of failure could be risky as well. It could give the developer the impression that the whole operation has failed and that no model has been created. They should be able to immediately retrieve the model id and delete it if desired. The current implementation allows that.

## Goals

- Discuss with the team what the expected behavior should be.
- If necessary, update the client code so it can behave as expected.
- Create tests to validate the chosen approach.
- Make sure existing tests still pass reliably.

### PR DESCRIPTION
**Pull request by maririos**
Fixes: https://github.com/Azure/azure-sdk-for-net/issues/11473

Follows the guidance:
- Throw exception if training method returns model with "invalid" status
- Exception message should include model_id, along with message from service that explains why it failed
- Include in documentation that a model is still created even if the training operation fails

For other operations, when `OperationStatus.Failed` we throw exception and if the user tries to access the `.Value` of the LRO, we throw an `InvalidOperationException` (similar to how it is handled in [KV ](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/keyvault/Azure.Security.KeyVault.Certificates/src/CertificateOperation.cs#L66)).

### CODE REVIEW

File: sdk/formrecognizer/Azure.AI.FormRecognizer/CHANGELOG.md [hunk 1]
```diff
@@ -28,6 +28,10 @@
 - The order of the values for `USReceiptType` have changed so that `Other` has now a value of `1`.
 - Parameter `useTrainingLabels` is now required for `FormTrainingClient.StartTraining`.
 - Protected constructors have been removed from `Operation` types, such as `TrainingOperation` or `RecognizeContentOperation`.
+- An `InvalidOperationException` is now raised if trying to access the `Value` property of a `TrainingOperation` when a trained model is invalid.
+- A `RequestFailedException` is now raised if a model with `status=="invalid"` is returned from the `StartTraining` and `StartTrainingAsync` methods.
+- A `RequestFailedException` is now raised if an operation like `StartRecognizeReceipts`, `StartRecognizeContent`, or `StartRecognizeCustomForms` fails.
+- An `InvalidOperationException` is now raised if trying to access the `Value` property of a `xxOperation` object when a the executed operation failed.
```
Comment (kristapratico) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/CHANGELOG.md:34]:
```suggestion
- An `InvalidOperationException` is now raised if trying to access the `Value` property of a `xxOperation` object when the executed operation failed.
```

**Review by kristapratico — APPROVED** [review]
LGTM

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs [hunk 2]
```diff
@@ -45,7 +45,18 @@ public class CopyModelOperation : Operation<CustomFormModelInfo>
...
 
         /// <inheritdoc/>
-        public override CustomFormModelInfo Value => OperationHelpers.GetValue(ref _value);
+        public override CustomFormModelInfo Value
+        {
+            get
+            {
+                if (HasCompleted && !HasValue)
+#pragma warning disable CA1065 // Do not raise exceptions in unexpected locations
+                    throw new InvalidOperationException("Copy model operation failed");
```
Comment (samvaity) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs:54]: We don't do this in Java.
Is it something only applicable for .NET or should other languages consider to follow?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeReceiptsOperation.cs [hunk 3]
```diff
@@ -48,22 +62,25 @@ public class RecognizeReceiptsOperation : Operation<RecognizedReceiptCollection>
         /// </summary>
         /// <param name="operationId">The ID of this operation.</param>
         /// <param name="client">The client used to check for completion.</param>
-        public RecognizeReceiptsOperation(string operationId, FormRecognizerClient client)
+        public RecognizeReceiptsOperation(string operationId,FormRecognizerClient client)
```
Comment (samvaity) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeReceiptsOperation.cs:65]: spacing

**Review by samvaity — APPROVED** [review]
LGTM :)

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs [hunk 4]
```diff
@@ -45,7 +45,18 @@ public class CopyModelOperation : Operation<CustomFormModelInfo>
         public override string Id { get; }
 
         /// <inheritdoc/>
-        public override CustomFormModelInfo Value => OperationHelpers.GetValue(ref _value);
+        public override CustomFormModelInfo Value
+        {
+            get
+            {
+                if (HasCompleted && !HasValue)
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs:52]: We have a race condition here.
```cs
if (update.Value.Status == OperationStatus.Succeeded)
{
  _hasCompleted = true;
  _value = ConvertValue(update.Value, _targetModelId, CustomFormModelStatus.Ready);
}
```
If you call this method between the attributions of `_hasCompleted` and `_value`, you'll get an `InvalidOperationException` with a `Copy model operation failed` message, which is not accurate.

Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs:52]: Maybe if you invert it:
```cs
if (update.Value.Status == OperationStatus.Succeeded)
{
  _value = ConvertValue(update.Value, _targetModelId, CustomFormModelStatus.Ready);
  _hasCompleted = true;
}
```
...?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs [hunk 5]
```diff
@@ -128,5 +151,36 @@ private static FormPageCollection ConvertValue(IReadOnlyList<PageResult_internal
 
             return new FormPageCollection(pages);
         }
+
+        private async ValueTask<RequestFailedException> CreateExceptionForFailedOperationAsync(bool async, IReadOnlyList<FormRecognizerError> errors)
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs:155]: It could be worth putting this method somewhere else so we don't repeat code. Error handling could change or be improved at some point in the future, so we could end up missing changes. We already have an `OperationExtensions` class, so it would be a good place to put it as an `internal static` method.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs [hunk 6]
```diff
@@ -128,5 +151,36 @@ private static FormPageCollection ConvertValue(IReadOnlyList<PageResult_internal
...
+            {
+                errorMessage = "RecognizeContent operation failed.";
+            }
+
+            var errorInfo = new Dictionary<string, string>();
+            int index = 0;
+
+            foreach (var error in errors)
+            {
+                errorInfo.Add($"error-{index}", $"{error.ErrorCode}: {error.Message}");
```
Comment (samvaity) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs:177]: Maybe I am missing something here, but unsure how the map would help here? The user would still have to iterate through it like it were a list. Or maybe I am

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs [hunk 7]
```diff
@@ -42,7 +42,18 @@ public class RecognizeCustomFormsOperation : Operation<RecognizedFormCollection>
...
 
         /// <inheritdoc/>
-        public override RecognizedFormCollection Value => OperationHelpers.GetValue(ref _value);
+        public override RecognizedFormCollection Value
+        {
+            get
+            {
+                if (HasCompleted && !HasValue)
+#pragma warning disable CA1065 // Do not raise exceptions in unexpected locations
+                    throw new InvalidOperationException("RecognizeCustomForms operation failed.");
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs:51]: @pakrym, is this the correct exception to raise here?  If we're doing this elsewhere in the SDK, would it make sense to pull this functionality into OperationHelpers in Core?

Comment (pakrym) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs:51]: It might be worth it if we find a design that works for multiple SDKs

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs (hunk 6 above)
Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs:177]: The map is something we pass to `CreateRequestFailedException` so it can list the errors in our exception message. It's not an object we pass to the user directly.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/TrainingOperation.cs [hunk 8]
```diff
@@ -96,16 +114,41 @@ private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToke
...
 
-                if (update.Value.ModelInfo.Status != CustomFormModelStatus.Training)
+                _response = update.GetRawResponse();
+
+                if (update.Value.ModelInfo.Status == CustomFormModelStatus.Ready)
                 {
                     _hasCompleted = true;
                     _value = new CustomFormModel(update.Value);
                 }
+                else if (update.Value.ModelInfo.Status == CustomFormModelStatus.Invalid)
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/TrainingOperation.cs:124]: If training fails, we are not letting users access the model and grab information like `RequestedOn`, `CompletedOn`, `TrainingDocuments` or `Errors`, right? The errors list can still be found in the exception message, though. Pointing it out in case we missed that, but I think we are not properly giving users the info the service has returned here.

(this argument doesn't make sense if this info is not returned in case of training failure, but I'm guessing it's not the case)

Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/TrainingOperation.cs:124]: @annelo-msft Thoughts?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs (hunk 7 above)
Comment (maririos) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs:51]: I looked into our guidelines and .NET guidelines when making this change and from my understanding `InvalidOperationException` will be better as it represents that the model is not right and therefore the user can't access the data.
When the user wants the value directly from the GET method, then it makes sense to raise `RequestFailedException`.

As for making a change that will be generic and benefit other libraries, I could only find some examples in keyvault and even in the library they have different behaviors per Operation class. It depends on the "state" the service is, and because that seems too specific, I decided to only leave this in FR.

Happy to look into it more if needed

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/TrainingOperation.cs (hunk 8 above)
Comment (maririos) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/TrainingOperation.cs:124]: This approach follows the recommendation from our architects.
When training fails, we are communicating with our exceptions the modelId so the user can go and 1) ask for more information with `GetCustomModel` 2) delete the model as it takes space :P

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs (hunk 5 above)
Comment (maririos) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs:155]: I thought about it, but wanted to keep it per operation in case in the future an endpoint decides to either add more to the errors, change structure, or the behavior change and we need to do some customization around it. What do you think?

Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs:155]: I think your concern is reasonable, but I wonder what's the likelihood of changes like that happening in the near future (specially because we don't expect big structure changes after GA). Even in that scenario, I think we should have the shared logic in a single place, and the customized bits hidden in their corresponding types (it depends a lot on the kind of change we will have, though).

I don't feel too strongly about it, though, so I wouldn't call it blocking if you think that keeping them separated may be worth it.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs (hunk 7 above)
Comment (annelo-msft) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs:51]: That sounds good.  Thanks for the investigation @maririos!

File: sdk/formrecognizer/Azure.AI.FormRecognizer/CHANGELOG.md [hunk 9]
```diff
@@ -28,6 +28,10 @@
 - The order of the values for `USReceiptType` have changed so that `Other` has now a value of `1`.
 - Parameter `useTrainingLabels` is now required for `FormTrainingClient.StartTraining`.
 - Protected constructors have been removed from `Operation` types, such as `TrainingOperation` or `RecognizeContentOperation`.
+- An `InvalidOperationException` is now raised if trying to access the `Value` property of a `TrainingOperation` when a trained model is invalid.
+- A `RequestFailedException` is now raised if a model with `status=="invalid"` is returned from the `StartTraining` and `StartTrainingAsync` methods.
+- A `RequestFailedException` is now raised if an operation like `StartRecognizeReceipts`, `StartRecognizeContent`, or `StartRecognizeCustomForms` fails.
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/CHANGELOG.md:33]: nit: this was already true for `StartRecognizeCustomForms`, wasn't it?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs [hunk 10]
```diff
@@ -111,8 +111,11 @@ public FormTrainingClient(Uri endpoint, TokenCredential credential, FormRecogniz
         /// <param name="useTrainingLabels">If <c>true</c>, use a label file created in the &lt;link-to-label-tool-doc&gt; to provide training-time labels for training a model. If <c>false</c>, the model will be trained from forms only.</param>
         /// <param name="trainingFileFilter">Filter to apply to the documents in the source path for training.</param>
         /// <param name="cancellationToken">A <see cref="CancellationToken"/> controlling the request lifetime.</param>
-        /// <returns>A <see cref="TrainingOperation"/> to wait on this long-running operation.  Its <see cref="TrainingOperation"/>.Value upon successful
-        /// completion will contain meta-data about the trained model.</returns>
+        /// <returns>
+        /// <para>A <see cref="TrainingOperation"/> to wait on this long-running operation. Its <see cref="TrainingOperation"/>.Value upon successful
+        /// completion will contain meta-data about the trained model.</para>
+        /// <para>If the training fails, an exception is raised, and a model with an "invalid" status is still created in the Form Recognizer resource.</para>
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormTrainingClient.cs:117]: Could we reword this to say something like:

- Even if training fails, a model is created in the Form Recognizer account with an "invalid" status.  A [type of]Exception will be raised containing the modelId to access this invalid model.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs [hunk 11]
```diff
@@ -101,14 +118,20 @@ private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToke
                     ? await _serviceClient.GetAnalyzeLayoutResultAsync(new Guid(Id), cancellationToken).ConfigureAwait(false)
                     : _serviceClient.GetAnalyzeLayoutResult(new Guid(Id), cancellationToken);
 
-                if (update.Value.Status == OperationStatus.Succeeded || update.Value.Status == OperationStatus.Failed)
+                _response = update.GetRawResponse();
+
+                if (update.Value.Status == OperationStatus.Succeeded)
                 {
                     _hasCompleted = true;
```
Comment (annelo-msft) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs:125]: Per @kinelski's comment, let's switch the order of `_hasCompleted` and `_value= ...` here and elsewhere.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs (hunk 7 above)
Comment (pakrym) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeCustomFormsOperation.cs:51]: Can we provide more diagnostics with this exception? `RecognizeCustomForms operation failed` is not very actionable.

Throwing RequestFailedException would allow us to at least include the final service response.

**Review by annelo-msft — APPROVED** [review]
Pending resolution of comments.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/TrainingOperation.cs [hunk 12]
```diff
@@ -96,16 +114,41 @@ private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToke
...
+                    throw await CreateExceptionForFailedOperationAsync(async, update.Value.TrainResult.Errors).ConfigureAwait(false);
+                }
             }
 
             return GetRawResponse();
         }
+
+        private async ValueTask<RequestFailedException> CreateExceptionForFailedOperationAsync(bool async, IReadOnlyList<FormRecognizerError> errors)
+        {
+            string errorMessage = $"Invalid model created with ID {_modelId}";
```
[the remark this answers is not in the document]
Comment (maririos) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/TrainingOperation.cs:138]: >Is this because we want to create our own exception here with the modelId?

Correct. We want to let the user know the modelid so they can access the model and take action on it.
As an example of error, if the bloburl provided is wrong, the message will say something like "Invalid container path" but a model was still created

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs (hunk 5 above)
Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/RecognizeContentOperation.cs:155]: Maybe using the `ClientCommon` class for that? We're only using it for validating the model id, for now. If `ClientCommon` doesn't feel right, maybe renaming it?

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs [hunk 13]
```diff
@@ -153,14 +166,17 @@ private async ValueTask<Response> UpdateStatusAsync(bool async, CancellationToke
 
                 if (update.Value.Status == OperationStatus.Succeeded)
                 {
-                    _hasCompleted = true;
+                    // we need to first assign a vaue and then mark the operation as completed to avoid race conditions
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs:169]: nit: vaue => value

Comment (kinelski) [reply sdk/formrecognizer/Azure.AI.FormRecognizer/src/CopyModelOperation.cs:169]: Maybe being more specific about what race condition we're talking about? I'm afraid this information will be lost if we don't document it properly:
```
We need to first assign a value and then mark the operation as completed to avoid a race condition with the getter in Value
```
These comments applies to all `***Operation` types.

File: sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormRecognizerClient.cs [hunk 14]
```diff
@@ -133,7 +133,7 @@ public virtual RecognizeContentOperation StartRecognizeContent(Stream form, Reco
             FormContentType contentType = recognizeOptions.ContentType ?? DetectContentType(form, nameof(form));
 
             ResponseWithHeaders<ServiceAnalyzeLayoutAsyncHeaders> response =  ServiceClient.AnalyzeLayoutAsync(contentType, form, cancellationToken);
-            return new RecognizeContentOperation(ServiceClient, response.Headers.OperationLocation);
+            return new RecognizeContentOperation(ServiceClient, Diagnostics,response.Headers.OperationLocation);
```
Comment (kinelski) [inline sdk/formrecognizer/Azure.AI.FormRecognizer/src/FormRecognizerClient.cs:136]: nit: missing space after `Diagnostics,`
